### DISCUSSION TITLE
Second EntraCP claims provider

### DISCUSSION THREAD

**Discussion by KamilPacanek** [Yvand/EntraCP#221] [Q&A] [answered]
Would it be possible to add the second EntraCP claims provider by modifying the default `EntraCP` claims provider name in the source code? The reason why: I have to serve two different web applications hosted on the same SharePoint Server, each configured with different tenants.

Would that cause some malfunction in EntraCP solution? Is that a no-go?

**Comment by Yvand** [comment]
Yes it is possible and there is an easier way:
You can create your own blank SharePoint solution project, reference NuGet package https://www.nuget.org/packages/Yvand.EntraCP/ and inherit class Yvand.EntraClaimsProvider.EntraCP with a custom name.

It is in my todo list to build a sample project to demonstrate how to do that, but you may give it a try if you are willing to.

**Reply by KamilPacanek** [reply]
Is it really enough? From my analysis over two days, I've come to conclusion that I had to modify far more details:

- change paths for hive deployable files like features definitions, admin and layouts pages
- change SP Administration links to add second category for separate configuration

Maybe more (like feature and package IDs) but those are because I was working on the same solution. Have you tried your approach from the beginning to the end?

**Reply by Yvand** [reply]
- You need to create your own features, at least one for the new claims provider
- The administration pages are not mandatory, EntraCP (or a claims provider that inherits it) can be configured entirely using PowerShell
- Each administration page has its logic in a .ascs, and has 3 properties (ClaimsProviderName, ConfigurationName and ConfigurationID) that can be set in the .aspx file that calls it. At minimum you need to create your own .aspx page to call the .ascs, but I believe you will also need to copy the .ascx file (just the markup, not the code behind). Tbh I need to do it before I can be sure if it is required

It is possible I need to update things in the NuGet package to ensure it can be used properly, as I mentioned I need to take the time to do it

**Reply by KamilPacanek** [reply]
Sure, that's fine by me for now because my approach looks like it should work, especially as you haven't mentioned any technical limitations for having two EntraCPs in coexistence. So far, I have two separate Admin sections and definitely original EntraCP works without any troubles - second one I'm waiting for Azure admins to set up theirs stuff, so I can validate that one as well.

By the way, I'm truly grateful for your project, it makes things much, much easier!

**Reply by Yvand** [reply, marked as answer]
Thank you for your nice feedback.
I guarantee that it is possible to have 2 EntraCP in same farm (one inheritng EntraCP class)
One thing you must really keep in mind though: Both wsp solutions reference/deploy the same dependent assemblies, so you need to be careful with update operations on wsp solutions.

On my side I need to actively work on creating clones of EntraCP.
Reason I did not do it yet is that I did not want to delay publishing EntraCP and I don't think there are many people who actually need this.
But I am committed to document it and improve the project/NuGet package if necessary.

**Reply by KamilPacanek** [reply]
Yup, dependencies got me already :P. But a quick update for original EntraCP was enough (and `machine.config` of course).

**Reply by Yvand** [reply]
@KamilPacanek sorry it took me so long, but I just added a sample project of a claims provider that inherits EntraCP.
You can find it here: https://github.com/Yvand/EntraCP/tree/dev/samples/CustomClaimsProvider

**Reply by KamilPacanek** [reply]
@Yvand not an issue at all. Thank you for the sample.

**Comment by KamilPacanek** [comment]
@Yvand :) I wanted to try that approach with inheriting/extending the `EntraCP` class.. but I noticed that there will be configuration issues for sure, right? Entra configuration is persisted and retrieved using GUID which is constant (`ClaimsProviderConstants`) and that's not taken into consideration here.
And I'm pretty sure I have to duplicate the EntraCP Admin section as well.

**Reply by Yvand** [reply]
The idea is that the custom class specifies its own configuration, as done in https://github.com/Yvand/EntraCP/blob/6a765034c6c1e659d11212787062458d58dfce6e/custom-claims-provider-samples/CustomClaimsProvider/EntraCP_Custom.cs#L23
In this simple example, it is hardcoded. It would be possible to store it as a persisted object but I did not make the effort to try it
